File: backend/src/main/java/org/opensource/smartair/services/QuantumLeapClient.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @Project smart-air-ngsi-ld
 * @Authors 
 *    - TT (trungthanhcva2206@gmail.com)
 *    - Tankchoi (tadzltv22082004@gmail.com)
 *    - Panh (panh812004.apn@gmail.com)
 * @Copyright (C) 2025 TAA. All rights reserved
 * @GitHub https://github.com/trungthanhcva2206/smart-air-ngsi-ld
 */
/*
 * Service for managing SSE streams (with historical data support)
 */
/*
 * Service to query historical data from QuantumLeap
 */
package org.opensource.smartair.services;

import io.netty.channel.ChannelOption;
import io.netty.handler.timeout.ReadTimeoutHandler;
import io.netty.handler.timeout.WriteTimeoutHandler;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.client.reactive.ReactorClientHttpConnector;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.client.WebClientResponseException;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.netty.http.client.HttpClient;
import reactor.netty.resources.ConnectionProvider;
import reactor.util.retry.Retry;

import java.time.Duration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

@Slf4j
@Service
public class QuantumLeapClient {

    private final WebClient webClient;
    private final String fiwareService;

    @Value("${quantumleap.query.lastN:720}")
    private int lastN;

    public QuantumLeapClient(
            @Value("${quantumleap.url}") String quantumLeapUrl,
            @Value("${quantumleap.fiware-service:hanoi}") String fiwareService) {

        this.fiwareService = fiwareService;

        // ✅ Cấu hình Connection Pool
        ConnectionProvider connectionProvider = ConnectionProvider.builder("quantumleap-pool")
                .maxConnections(50)
                .maxIdleTime(Duration.ofSeconds(20))
                .maxLifeTime(Duration.ofSeconds(60))
                .pendingAcquireTimeout(Duration.ofSeconds(45))
                .evictInBackground(Duration.ofSeconds(120))
                .build();

        // ✅ Cấu hình HttpClient với timeouts
        HttpClient httpClient = HttpClient.create(connectionProvider)
                .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 10000)
                .responseTimeout(Duration.ofSeconds(30))
                .doOnConnected(conn -> conn
                        .addHandlerLast(new ReadTimeoutHandler(30, TimeUnit.SECONDS))
                        .addHandlerLast(new WriteTimeoutHandler(30, TimeUnit.SECONDS)));

        // ✅ Tạo WebClient với custom HttpClient
        this.webClient = WebClient.builder()
                .baseUrl(quantumLeapUrl)
                .clientConnector(new ReactorClientHttpConnector(httpClient))
                .build();

        log.info("✅ QuantumLeap WebClient initialized: {}", quantumLeapUrl);
    }

    // ============ Weather History ============
    public Mono<Map<String, Object>> getAggregatedAirQualityHistory(List<String> districts) {
        log.info("📊 Fetching aggregated air quality history for {} districts", districts.size());

        // Fetch history for each district in parallel
        List<Mono<Map.Entry<String, Object>>> monos = districts.stream()
                .map(district -> {
                    String entityId = String.format("urn:ngsi-ld:AirQualityObserved:Hanoi-%s", district);

                    return getAirQualityHistory(district)
                            .map(historyData -> Map.entry(district, (Object) historyData))
                            .onErrorResume(error -> {
                                log.warn("⚠️ Failed to fetch history for {}: {}", district, error.getMessage());
                                return Mono.just(Map.entry(district, (Object) new HashMap<>()));
                            });
                })
                .toList();

        // Combine all results into a single Map
        return Flux.merge(monos)
                .collectMap(Map.Entry::getKey, Map.Entry::getValue)
                .doOnSuccess(aggregated -> log.info("✅ Successfully aggregated history for {}/{} districts",
                        aggregated.size(), districts.size()))
                .doOnError(error -> log.error("❌ Error aggregating history: {}", error.getMessage()));
    }

    /**
     * ✅ NEW: Fetch aggregated weather history for ALL districts
     */
    public Mono<Map<String, Object>> getAggregatedWeatherHistory(List<String> districts) {
        log.info("📊 Fetching aggregated weather history for {} districts", districts.size());

        List<Mono<Map.Entry<String, Object>>> monos = districts.stream()
                .map(district -> getWeatherHistory(district)
                        .map(historyData -> Map.entry(district, (Object) historyData))
                        .onErrorResume(error -> {
                            log.warn("⚠️ Failed to fetch weather history for {}: {}", district, error.getMessage());
                            return Mono.just(Map.entry(district, (Object) new HashMap<>()));
                        }))
                .toList();

        return Flux.merge(monos)
                .collectMap(Map.Entry::getKey, Map.Entry::getValue)
                .doOnSuccess(aggregated -> log.info("✅ Successfully aggregated weather history for {}/{} districts",
                        aggregated.size(), districts.size()));
    }

    public Mono<Map<String, Object>> getWeatherHistory(String district) {
        String entityId = buildWeatherEntityId(district);

        log.info("Querying QuantumLeap for entity: {}", entityId);

        return webClient.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/v2/entities/{entityId}")
                        .queryParam("lastN", lastN)
                        .queryParam("type", "WeatherObserved")
                        .build(entityId))
                .header("Fiware-Service", fiwareService)
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {
                }) // ✅ FIX: Use ParameterizedTypeReference
                .retryWhen(Retry.backoff(3, Duration.ofSeconds(1))
                        .filter(throwable -> throwable instanceof org.springframework.web.reactive.function.client.WebClientRequestException)
                        .doBeforeRetry(retrySignal -> log.warn("🔄 Retrying QuantumLeap request for {} (attempt {})",
                                entityId, retrySignal.totalRetries() + 1)))
                .doOnSuccess(data -> log.info("✅ Successfully fetched historical data for: {}", entityId))
                .onErrorResume(WebClientResponseException.NotFound.class, e -> {
                    log.warn("⚠️ No historical data found in QuantumLeap for: {}", entityId);
                    return Mono.just(Map.of());
                })
                .onErrorResume(WebClientResponseException.class, e -> {
                    log.error("❌ HTTP error fetching data from QuantumLeap for {}: {} - {}",
                            entityId, e.getStatusCode(), e.getMessage());
                    return Mono.just(Map.of());
                })
                .onErrorResume(Exception.class, e -> {
                    log.error("❌ Error fetching historical data from QuantumLeap for: {}", entityId, e);
                    return Mono.just(Map.of());
                })
                .doOnTerminate(() -> log.debug("✅ QuantumLeap request completed for: {}", entityId))
                .defaultIfEmpty(Map.of());
    }

    // ============ Air Quality History ============

    public Mono<Map<String, Object>> getAirQualityHistory(String district) {
        String entityId = buildAirQualityEntityId(district);

        log.info("Querying QuantumLeap for entity: {}", entityId);

        return webClient.get()
                .uri(uriBuilder -> uriBuilder
                        .path("/v2/entities/{entityId}")
                        .queryParam("lastN", lastN)
                        .queryParam("type", "AirQualityObserved")
                        .build(entityId))
                .header("Fiware-Service", fiwareService)
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {
                }) // ✅ FIX: Use ParameterizedTypeReference
                .retryWhen(Retry.backoff(3, Duration.ofSeconds(1))
                        .filter(throwable -> throwable instanceof org.springframework.web.reactive.function.client.WebClientRequestException)
                        .doBeforeRetry(retrySignal -> log.warn("🔄 Retrying QuantumLeap request for {} (attempt {})",
                                entityId, retrySignal.totalRetries() + 1)))
                .doOnSuccess(data -> log.info("✅ Successfully fetched historical data for: {}", entityId))
                .onErrorResume(WebClientResponseException.NotFound.class, e -> {
                    log.warn("⚠️ No historical data found in QuantumLeap for: {}", entityId);
                    return Mono.just(Map.of());
                })
                .onErrorResume(WebClientResponseException.class, e -> {
                    log.error("❌ HTTP error fetching data from QuantumLeap for {}: {} - {}",
                            entityId, e.getStatusCode(), e.getMessage());
                    return Mono.just(Map.of());
                })
                .onErrorResume(Exception.class, e -> {
                    log.error("❌ Error fetching historical data from QuantumLeap for: {}", entityId, e);
                    return Mono.just(Map.of());
                })
                .doOnTerminate(() -> log.debug("✅ QuantumLeap request completed for: {}", entityId))
                .defaultIfEmpty(Map.of());
    }

    // ============ Attribute History (Generic) ============

    /**
     * Get historical data for a specific attribute from QuantumLeap
     * Generic method for any entity type (Weather, AirQuality, Device)
     * 
     * @param entityId   Full URN (e.g.,
     *                   "urn:ngsi-ld:WeatherObserved:Hanoi-PhuongBaDinh")
     * @param attrName   Attribute name (e.g., "temperature", "pm2_5")
     * @param fromDate   Start date ISO 8601 (optional)
     * @param toDate     End date ISO 8601 (optional)
     * @param aggrMethod Aggregation: avg, sum, min, max, count (optional)
     * @param aggrPeriod Period: hour, day, week, month (optional)
     * @param lastN      Number of points (optional, fallback if no dates)
     * @return Mono with historical data
     */
    public Mono<Map<String, Object>> getAttributeHistory(
            String entityId, String attrName, String fromDate, String toDate,
            String aggrMethod, String aggrPeriod, Integer lastN) {

        // Clean and validate input parameters
        String cleanFromDate = fromDate != null ? fromDate.trim().replaceAll("^['\"]|['\"]$", "") : null;
        String cleanToDate = toDate != null ? toDate.trim().replaceAll("^['\"]|['\"]$", "") : null;

        log.info(
                "Querying QuantumLeap: {} attr: {} (fromDate: {}, toDate: {}, aggrMethod: {}, aggrPeriod: {}, lastN: {})",
                entityId, attrName, cleanFromDate, cleanToDate, aggrMethod, aggrPeriod, lastN);

        return webClient.get()
                .uri(uriBuilder -> {
                    var builder = uriBuilder
                            .path("/v2/entities/{entityId}/attrs/{attrName}");

                    // QuantumLeap parameter order: aggrMethod -> aggrPeriod -> fromDate -> toDate
                    if (aggrMethod != null) {
                        builder.queryParam("aggrMethod", aggrMethod);
                    }
                    if (aggrPeriod != null) {
                        builder.queryParam("aggrPeriod", aggrPeriod);
                    }
                    if (cleanFromDate != null && cleanToDate != null) {
                        builder.queryParam("fromDate", cleanFromDate);
                        builder.queryParam("toDate", cleanToDate);
                    } else if (lastN != null) {
                        builder.queryParam("lastN", lastN);
                    } else {
                        builder.queryParam("lastN", this.lastN);
                    }

                    // Spring auto URL-encodes path variables
                    return builder.build(entityId, attrName);
                })
                .header("Fiware-Service", fiwareService)
                .retrieve()
                .bodyToMono(new ParameterizedTypeReference<Map<String, Object>>() {
                })
                .retryWhen(Retry.backoff(3, Duration.ofSeconds(1))
                        .filter(throwable -> throwable instanceof org.springframework.web.reactive.function.client.WebClientRequestException)
                        .doBeforeRetry(
                                retrySignal -> log.warn("Retrying QuantumLeap request for {} attr {} (attempt {})",
                                        entityId, attrName, retrySignal.totalRetries() + 1)))
                .doOnSuccess(data -> log.info("Successfully fetched {} history for: {}", attrName, entityId))
                .onErrorResume(WebClientResponseException.NotFound.class, e -> {
                    log.warn("No historical data found in QuantumLeap for: {} attr: {}", entityId, attrName);
                    return Mono.just(Map.of());
                })
                .onErrorResume(WebClientResponseException.class, e -> {
                    log.error("HTTP error fetching {} history from QuantumLeap for {}: {} - {}",
                            attrName, entityId, e.getStatusCode(), e.getMessage());
                    return Mono.just(Map.of());
                })
                .onErrorResume(Exception.class, e -> {
                    log.error("Error fetching {} history from QuantumLeap for: {}", attrName, entityId, e);
                    return Mono.just(Map.of());
                })
                .doOnTerminate(() -> log.debug("QuantumLeap request completed for: {} attr: {}", entityId, attrName))
                .defaultIfEmpty(Map.of());
    }

    /**
     * Get weather attribute history by district name
     * Convenience method that builds WeatherObserved URN
     */
    public Mono<Map<String, Object>> getWeatherAttributeHistory(
            String district, String attrName, String fromDate, String toDate,
            String aggrMethod, String aggrPeriod, Integer lastN) {

        String entityId = buildWeatherEntityId(district);
        return getAttributeHistory(entityId, attrName, fromDate, toDate, aggrMethod, aggrPeriod, lastN);
    }

    /**
     * Get air quality attribute history by district name
     * Convenience method that builds AirQualityObserved URN
     */
    public Mono<Map<String, Object>> getAirQualityAttributeHistory(
            String district, String attrName, String fromDate, String toDate,
            String aggrMethod, String aggrPeriod, Integer lastN) {

        String entityId = buildAirQualityEntityId(district);
        return getAttributeHistory(entityId, attrName, fromDate, toDate, aggrMethod, aggrPeriod, lastN);
    }

    // ============ Helper Methods ============

    private String buildWeatherEntityId(String district) {
        return String.format("urn:ngsi-ld:WeatherObserved:Hanoi-%s", district);
    }

    private String buildAirQualityEntityId(String district) {
        return String.format("urn:ngsi-ld:AirQualityObserved:Hanoi-%s", district);
    }

}

File: etl/Core_ETL/quantum_leap_client.py
"""
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @Project smart-air-ngsi-ld
 * @Authors 
 *    - TT (trungthanhcva2206@gmail.com)
 *    - Tankchoi (tadzltv22082004@gmail.com)
 *    - Panh (panh812004.apn@gmail.com)
 * @Copyright (C) 2025 CHK. All rights reserved
 * @GitHub https://github.com/trungthanhcva2206/smart-air-ngsi-ld
 */
"""

"""
QuantumLeap Client for Time Series Data Storage
"""
import requests
import logging
from typing import Dict, List, Optional
from datetime import datetime, timedelta
from .config import QUANTUMLEAP_EXTERNAL_URL, ORION_LD_TENANT

logger = logging.getLogger(__name__)


class QuantumLeapClient:
    """Client for QuantumLeap Time Series API"""
    
    def __init__(self):
        self.base_url = QUANTUMLEAP_EXTERNAL_URL 
        self.tenant = ORION_LD_TENANT
        
        # QuantumLeap headers
        self.headers = {
            'Content-Type': 'application/json',
            'Fiware-Service': self.tenant.lower()
        }
    
    def notify_entity(self, entity: Dict) -> bool:
        """
        Send entity to QuantumLeap for time series storage
        This is typically done via subscription, but can be done manually
        
        Args:
            entity: NGSI-LD entity dictionary
            
        Returns:
            True if successful, False otherwise
        """
        try:
            # QuantumLeap expects v2 notify format
            # Convert NGSI-LD to v2 format
            v2_entity = self._convert_to_v2(entity)
            
            url = f"{self.base_url}/v2/notify"
            payload = {
                "data": [v2_entity],
                "subscriptionId": "manual-notification"
            }
            
            response = requests.post(
                url,
                json=payload,
                headers=self.headers,
                timeout=10
            )
            
            if response.status_code in [200, 201, 204]:
                logger.info(f"Successfully notified QuantumLeap for entity: {entity['id']}")
                return True
            else:
                logger.error(
                    f"Error notifying QuantumLeap for {entity['id']}: "
                    f"{response.status_code} - {response.text}"
                )
                return False
                
        except Exception as e:
            logger.error(f"Error notifying QuantumLeap: {e}")
            return False
    
    def get_entity_history(self, entity_id: str, entity_type: str,
                          from_date: Optional[str] = None,
                          to_date: Optional[str] = None,
                          limit: int = 100,
                          offset: int = 0) -> Optional[Dict]:
        """
        Get historical data for an entity
        
        Args:
            entity_id: Entity ID
            entity_type: Entity type
            from_date: Start date (ISO 8601)
            to_date: End date (ISO 8601)
            limit: Maximum number of records
            offset: Offset for pagination
            
        Returns:
            Historical data or None if request fails
        """
        try:
            url = f"{self.base_url}/v2/entities/{entity_id}"
            params = {
                'type': entity_type,
                'limit': limit,
                'offset': offset
            }
            
            if from_date:
                params['fromDate'] = from_date
            if to_date:
                params['toDate'] = to_date
            
            response = requests.get(
                url,
                params=params,
                headers=self.headers,
                timeout=10
            )
            
            if response.status_code == 200:
                logger.info(f"Successfully retrieved history for {entity_id}")
                return response.json()
            else:
                logger.error(
                    f"Error getting history for {entity_id}: "
                    f"{response.status_code} - {response.text}"
                )
                return None
                
        except Exception as e:
            logger.error(f"Error getting entity history: {e}")
            return None
    
    def get_attribute_history(self, entity_id: str, entity_type: str,
                             attr_name: str,
                             from_date: Optional[str] = None,
                             to_date: Optional[str] = None,
                             limit: int = 100,
                             offset: int = 0,
                             last_n: Optional[int] = None) -> Optional[Dict]:
        """
        Get historical data for a specific attribute
        
        Args:
            entity_id: Entity ID
            entity_type: Entity type
            attr_name: Attribute name
            from_date: Start date (ISO 8601)
            to_date: End date (ISO 8601)
            limit: Maximum number of records
            offset: Offset for pagination
            last_n: Get last N values
            
        Returns:
            Attribute historical data or None if request fails
        """
        try:
            url = f"{self.base_url}/v2/entities/{entity_id}/attrs/{attr_name}"
            params = {
                'type': entity_type,
                'limit': limit,
                'offset': offset
            }
            
            if from_date:
                params['fromDate'] = from_date
            if to_date:
                params['toDate'] = to_date
            if last_n:
                params['lastN'] = last_n
            
            response = requests.get(
                url,
                params=params,
                headers=self.headers,
                timeout=10
            )
            
            if response.status_code == 200:
                logger.info(f"Successfully retrieved attribute history for {entity_id}.{attr_name}")
                return response.json()
            else:
                logger.error(
                    f"Error getting attribute history for {entity_id}.{attr_name}: "
                    f"{response.status_code} - {response.text}"
                )
                return None
                
        except Exception as e:
            logger.error(f"Error getting attribute history: {e}")
            return None
    
    def get_aggregated_data(self, entity_type: str,
                           attr_name: str,
                           aggr_method: str = 'avg',
                           aggr_period: str = 'hour',
                           from_date: Optional[str] = None,
                           to_date: Optional[str] = None) -> Optional[Dict]:
        """
        Get aggregated time series data
        
        Args:
            entity_type: Entity type
            attr_name: Attribute name
            aggr_method: Aggregation method (avg, sum, min, max, count)
            aggr_period: Aggregation period (second, minute, hour, day, month, year)
            from_date: Start date (ISO 8601)
            to_date: End date (ISO 8601)
            
        Returns:
            Aggregated data or None if request fails
        """
        try:
            url = f"{self.base_url}/v2/types/{entity_type}/attrs/{attr_name}"
            params = {
                'aggrMethod': aggr_method,
                'aggrPeriod': aggr_period
            }
            
            if from_date:
                params['fromDate'] = from_date
            if to_date:
                params['toDate'] = to_date
            
            response = requests.get(
                url,
                params=params,
                headers=self.headers,
                timeout=10
            )
            
            if response.status_code == 200:
                logger.info(
                    f"Successfully retrieved aggregated data for {entity_type}.{attr_name}"
                )
                return response.json()
            else:
                logger.error(
                    f"Error getting aggregated data: "
                    f"{response.status_code} - {response.text}"
                )
                return None
                
        except Exception as e:
            logger.error(f"Error getting aggregated data: {e}")
            return None
    
    def delete_entity_history(self, entity_id: str, entity_type: str) -> bool:
        """
        Delete all historical data for an entity
        
        Args:
            entity_id: Entity ID
            entity_type: Entity type
            
        Returns:
            True if successful, False otherwise
        """
        try:
            url = f"{self.base_url}/v2/entities/{entity_id}"
            params = {'type': entity_type}
            
            response = requests.delete(
                url,
                params=params,
                headers=self.headers,
                timeout=10
            )
            
            if response.status_code in [200, 204]:
                logger.info(f"Successfully deleted history for {entity_id}")
                return True
            else:
                logger.error(
                    f"Error deleting history for {entity_id}: "
                    f"{response.status_code} - {response.text}"
                )
                return False
                
        except Exception as e:
            logger.error(f"Error deleting entity history: {e}")
            return False
    
    def get_health(self) -> Optional[Dict]:
        """
        Check QuantumLeap health status
        
        Returns:
            Health status or None if request fails
        """
        try:
            url = f"{self.base_url}/health"
            response = requests.get(url, timeout=5)
            
            if response.status_code == 200:
                return response.json()
            else:
                return None
                
        except Exception as e:
            logger.error(f"Error checking QuantumLeap health: {e}")
            return None
    
    def get_version(self) -> Optional[Dict]:
        """
        Get QuantumLeap version information
        
        Returns:
            Version info or None if request fails
        """
        try:
            url = f"{self.base_url}/version"
            response = requests.get(url, timeout=5)
            
            if response.status_code == 200:
                return response.json()
            else:
                return None
                
        except Exception as e:
            logger.error(f"Error getting QuantumLeap version: {e}")
            return None
    
    def _convert_to_v2(self, ngsi_ld_entity: Dict) -> Dict:
        """
        Convert NGSI-LD entity to NGSIv2 format for QuantumLeap
        
        Args:
            ngsi_ld_entity: NGSI-LD entity
            
        Returns:
            NGSIv2 entity
        """
        v2_entity = {
            'id': ngsi_ld_entity['id'],
            'type': ngsi_ld_entity['type']
        }
        
        for key, value in ngsi_ld_entity.items():
            if key in ['id', 'type', '@context']:
                continue
            
            if isinstance(value, dict):
                if value.get('type') == 'Property':
                    v2_entity[key] = {
                        'value': value.get('value'),
                        'type': self._infer_v2_type(value.get('value'))
                    }
                    if 'observedAt' in value:
                        v2_entity[key]['metadata'] = {
                            'timestamp': {
                                'value': value['observedAt'],
                                'type': 'DateTime'
                            }
                        }
                    if 'unitCode' in value:
                        if 'metadata' not in v2_entity[key]:
                            v2_entity[key]['metadata'] = {}
                        v2_entity[key]['metadata']['unitCode'] = {
                            'value': value['unitCode'],
                            'type': 'Text'
                        }
                
                elif value.get('type') == 'GeoProperty':
                    v2_entity[key] = {
                        'value': value.get('value'),
                        'type': 'geo:json'
                    }
                
                elif value.get('type') == 'Relationship':
                    v2_entity[key] = {
                        'value': value.get('object'),
                        'type': 'Relationship'
                    }
        
        return v2_entity
    
    def _infer_v2_type(self, value) -> str:
        """Infer NGSIv2 type from value"""
        if isinstance(value, bool):
            return 'Boolean'
        elif isinstance(value, int):
            return 'Integer'
        elif isinstance(value, float):
            return 'Number'
        elif isinstance(value, str):
            return 'Text'
        elif isinstance(value, list):
            return 'Array'
        elif isinstance(value, dict):
            return 'StructuredValue'
        else:
            return 'Text'